code/modules/projectiles/guns/misc/blastcannon.dm

```diff
@@ -0,0 +1,131 @@
+	if(istype(O, /obj/item/transfer_valve))
+		var/obj/item/transfer_valve/T = O
+		if(!T.tank_one || !T.tank_two)
+			to_chat(user, "<span class='warning'>What good would an incomplete bomb do?</span>")
+			return FALSE
+		if(!user.drop_item())
+			to_chat(user, "<span class='warning'>[T] seems to be stuck to your hand!</span>")
+			return FALSE
+		user.visible_message("<span class='warning'>[user] attaches [T] to [src]!</span>")
+		T.forceMove(src)
+		bomb = T
+		update_icon()
+		return TRUE
+	return ..()
+
+/obj/item/gun/blastcannon/proc/calculate_bomb()
+	if(!istype(bomb)||!istype(bomb.tank_one)||!istype(bomb.tank_two))
+		return 0
+	var/datum/gas_mixture/temp = new()	//directional buff.
+	temp.volume = 60
+	temp.merge(bomb.tank_one.air_contents.remove_ratio(1))
+	temp.merge(bomb.tank_two.air_contents.remove_ratio(2))
+	for(var/i in 1 to 6)
+		temp.react()
+	var/pressure = temp.return_pressure()
+	qdel(temp)
+	if(pressure < TANK_FRAGMENT_PRESSURE)
+		return 0
+	return (pressure / TANK_FRAGMENT_SCALE)
+
+/obj/item/gun/blastcannon/afterattack(atom/target, mob/user, flag, params)
+	if((!bomb) || (!target) || (get_dist(get_turf(target), get_turf(user)) <= 2))
+		return ..()
+	var/power = calculate_bomb()
+	QDEL_NULL(bomb)
+	update_icon()
+	var/heavy = power * 0.2
+	var/medium = power * 0.5
+	var/light = power
+	user.visible_message("<span class='danger'>[user] opens [bomb] on \his [name] and fires a blast wave at [target]!</span>","<span class='danger'>You open [bomb] on your [name] and fire a blast wave at [target]!</span>")
+	playsound(user, "explosion", 100, 1)
+	var/turf/starting = get_turf(user)
+	var/turf/targturf = get_turf(target)
+	var/log_str = "Blast wave fired from [ADMIN_COORDJMP(starting)] ([get_area_name(user, TRUE)]) at [ADMIN_COORDJMP(targturf)] ([target.name]) by [user.name]([user.ckey]) with power [heavy]/[medium]/[light]."
+	message_admins(log_str)
+	log_game(log_str)
+	var/obj/item/projectile/blastwave/BW = new(loc, heavy, medium, light)
+	BW.preparePixelProjectile(target, get_turf(target), user, params, 0)
+	BW.fire()
+
+/obj/item/projectile/blastwave
+	name = "blast wave"
+	icon_state = "blastwave"
+	damage = 0
+	nodamage = FALSE
+	forcedodge = TRUE
+	range = 150
+	var/heavyr = 0
+	var/mediumr = 0
+	var/lightr = 0
+
+/obj/item/projectile/blastwave/New(loc, _h, _m, _l)
+	..()
+	heavyr = _h
+	mediumr = _m
+	lightr = _l
+
+/obj/item/projectile/blastwave/Range()
+	..()
+	var/amount_destruction = 0
+	if(heavyr)
+		amount_destruction = EXPLODE_DEVASTATE
+	else if(mediumr)
+		amount_destruction = EXPLODE_HEAVY
+	else if(lightr)
+		amount_destruction = EXPLODE_LIGHT
+	if(amount_destruction && isturf(loc))
+		var/turf/T = loc
+		for(var/thing in T.contents)
+			var/atom/AM = thing
+			if(AM && AM.simulated)
+				AM.ex_act(amount_destruction)
+				CHECK_TICK
+		T.ex_act(amount_destruction)
+	else
+		qdel(src)
+	heavyr = max(heavyr - 1, 0)
+	mediumr = max(mediumr - 1, 0)
+	lightr = max(lightr - 1, 0)
```

Comment:
I kept having problems with the projectile calling `ex_act(EXPLODE_DEVASTATE)` on everything; TG used `heavyr--`. The problem with this is that when the values go negative, it still counts as `if(heavyr)`, and continues to do horrendous damage until the projectile travels a full 150 turfs, then deletes. This rectifies the problem.